rfcs/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,288 @@
+`libjpeg` C libraries) that `panic` across C frames. The safety of such
+unwinding relies on compatibility between Rust's unwinding mechanism and the
+native exception mechanisms in GCC, LLVM, and MSVC.
+
+Additionally, there are libraries such as `rlua` that rely on `longjmp` across
+Rust frames; on Windows, `longjmp` is implemented via [forced
+unwinding](forced-unwinding). The current `rustc` implementation makes it safe
+to `longjmp` across Rust frames without `Drop` types, but this is not formally
+specified in an RFC or by the Reference.
+
+The desire for this feature has been previously discussed on other RFCs,
+including [#2699][rfc-2699] and [#2753][rfc-2753].
+
+[rfc-2699]: https://github.com/rust-lang/rfcs/pull/2699
+[rfc-2753]: https://github.com/rust-lang/rfcs/pull/2573
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring an external function that may unwind, such as an entrypoint to a
+C++ library, use `"C unwind"`:
+
+```
+extern "C unwind" {
+  fn may_throw();
+}
+```
+
+Rust functions that call a possibly-unwinding external function should either
+use the default Rust ABI (which can be made explicit with `extern "Rust"`) or
+the `"C unwind"` ABI:
+
+```
+extern "C unwind" fn can_unwind() {
+  // ...
+```
+
+Using the `"C unwind"` ABI to "sandwich" Rust frames between frames from
+another language (such as C++) allows an exception initiated in a callee frame
+in the other language to traverse the intermediate Rust frames before being
+caught in the caller frames. I.e., a C++ exception may be thrown,
+cross into Rust via an `extern "C unwind"` function declaration, safely unwind
+the Rust frames, and cross back into C++ (where it may be caught) via a Rust
+`"C unwind"` function definition.
+
+Conversely, languages that support the native unwinding mechanism, such as C++,
+may be "sandwiched" between Rust frames, so that Rust `panic`s may safely
+unwind the C++ frames, if the Rust code declares both the C++ entrypoint and
+the Rust entrypoint using `"C unwind"`.
+
+## Forced unwinding
+[forced-unwinding]: #forced-unwinding
+
+This is a special kind of unwinding used to implement `longjmp` on Windows and
+`pthread_exit` in `glibc`. A brief explanation is provided in [this Inside Rust
+blog post][inside-rust-forced]. This RFC distinguishes forced unwinding from
+other types of foreign unwinding.
+
+Since language features and library functions implemented using forced
+unwinding on some platforms use other mechanisms on other platforms, Rust code
+cannot rely on forced unwinding to invoke destructors (calling `drop` on `Drop`
+types). In other words, a forced unwind operation on one platform will simply
+deallocate Rust frames without true unwinding on other platforms.
+
+This RFC specifies that it is undefined behavior to cross Rust frames with
+destructors via such language features and library functions, regardless of the
+platform.
```

Comment:
This seems good to me. It's probably worth stating explicitly that this is true for both "C" and "C unwind" functions -- and also that it IS OK to unwind if there are no destructors.

Maybe something like:

"Forced unwinding" works differently than ordinary unwinding. For both "C" and "C unwind" functions, unwinding a Rust frame with forced unwinding is:

* *undefined behavior* if there are destructors in scope (or a `catch_unwind` call)
* *defined behavior* otherwise

As an example, this means that Rust code can invoke `longjmp` (or `C` functions which invoke `longjmp`) using the "C" ABI, and that longjmp can unwind Rust frames, so long as those frames do not contain any pending destructors or make use of `catch_unwind`. If those Rust frames *do* contain pending destructors, then invoking `longjmp` would be *undefined behavior* (and hence a bug).

Reply:
Adopted with some rephrasing.